File: graph/496-countPairs-unionFind.py

class DisjointSet(object):
    def __init__(self,n) -> None:
        self.par=list(range(n))
        self.size=[1]*n
        self.total=n
    
    def merge(self,a,b):
        a,b=self.find(a),self.find(b)
        if a==b:
            return False
        if self.size[a]<self.size[b]:
            a,b=b,a
        self.par[b]=a
        self.size[a]+=self.size[b]
        
        self.size[b]=0
        self.total-=1
        return True
    
    def find(self,a):
        if a !=self.par[a]:
            self.par[a]=self.find(self.par[a])
        return self.par[a]
    
    def getSize(self,a):
        return self.size[a]
    
    def count(self):
        return self.total
    
    
def countPair(n,edges):
    ds=DisjointSet(n)
    ans=0
    for a,b in edges:
        ds.merge(a,b)
    
    for i in range(n):
        c=ds.getSize(i)
        ans+=(n-c)*c
    return ans//2


n = 7
edges = [[0,2],[0,5],[2,4],[1,6],[5,4]]
print(countPair(n,edges))
    

File: graph/496-countPairs-unionFind.cpp
#include<iostream>
#include<vector>
#include<map>

using VEC=std::vector<int>;
using MATRIX=std::vector<VEC>;
using G=std::map<int,VEC>;


class DisjointSet
{
    private:
        VEC m_par{};
        VEC m_size{};
    
    public:
        
        DisjointSet(int n)
        {
            for(int i=0;i<n;++i)
            {
                m_par.push_back(i);
                m_size.push_back(1);
            }
        }
        int find(int i)
        {
            if(i!=m_par[i])
            {
                m_par[i]=find(m_par[i]);
            }
            return m_par[i];
        }
        bool merge(int a,int b)
        {
            int pa=find(a),pb=find(b);
            if(pa==pb)
                return false;
            if(m_size[a]<m_size[b])
            {
                int pTmp=pa;
                pa=pb;
                pb=pTmp;
            }
            m_par[pb]=pa;
            m_size[pa]+=m_size[pb];
            m_size[pb]=0;
        }
        int getSize(int a) const 
        {
            return m_size[a];
        }
};

int countPair(int n,const MATRIX& edges)
{
    G g{};

    DisjointSet ds{n};
    int ans=0;
    for(auto &v:edges)
    {
        ds.merge(v[0],v[1]);
    }
    for(int i=0;i<n;++i)
    {
        int c=ds.getSize(i);
        ans+=(n-c)*c;
    }
    return ans/2;
}

int main()
{
        int n = 7;
    MATRIX edges = {{0,2},{0,5},{2,4},{1,6},{5,4}};
    std::cout<<(countPair(n,edges));
}
